src/ProgressOnderwijsUtils/Collections/CachedTreeBuilder.cs

```diff
@@ -1,22 +1,22 @@
 namespace ProgressOnderwijsUtils.Collections;
 
-static class CachedTreeBuilder<TInput, TNodeValue>
+static class TreeBuilder<TInput, TOutput>
 {
     sealed class TreeNodeBuilder
     {
         public required TInput value;
         public TreeNodeBuilder? parent;
-        public int idxInParent;
-        public Tree<TNodeValue>[] kids = [];
+        public required int idxInParent;
+        public TOutput[] kids = [];
     }
 
     [Pure]
-    public static Tree<TNodeValue> Resolve(TInput rootNodeValue, Func<TInput, IEnumerable<TInput>?> kidLookup, Func<TInput, Tree<TNodeValue>[], Tree<TNodeValue>> map)
+    public static TOutput Build(TInput rootNodeValue, Func<TInput, IEnumerable<TInput>?> kidLookup, Func<TInput, TOutput[], TOutput> map)
     {
         var needsKids = new Stack<TreeNodeBuilder>();
 
         var generatedNodes = 0;
-        var rootBuilder = new TreeNodeBuilder { value = rootNodeValue, };
+        var rootBuilder = new TreeNodeBuilder { value = rootNodeValue, idxInParent = 0, };
```

Comment:
Does `required` force initialization of `idxInParent` here?

Reply:
yep!  It's not really important either way, but algorithms such as these are hard enough to follow as is and being explicit helped me here.  non-required, implicit init would be just as good, I'm sure.